File: Problem Solving/Algorithms/Implementation/Strong Password.py
#!/bin/python3

import math
import os
import random
import re
import sys

#
# Complete the 'minimumNumber' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER n
#  2. STRING password
#

def minimumNumber(n, password):
    # Return the minimum number of characters to make the password strong
    numbers = "0123456789"
    lower_case = "abcdefghijklmnopqrstuvwxyz"
    upper_case = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    special_characters = "!@#$%^&*()-+"

    has_numbers = False
    has_lower_case = False
    has_upper_case = False
    has_special_characters = False
    
    for char in password:
        if char in numbers:
            has_numbers = True
        elif char in lower_case:
            has_lower_case = True
        elif char in upper_case:
            has_upper_case = True
        elif char in special_characters:
            has_special_characters = True

    count = 0
    if has_numbers is False:
        count += 1
    if has_lower_case is False:
        count += 1
    if has_upper_case is False:
        count += 1
    if has_special_characters is False:
        count += 1

    if len(password) + count < 6:
        return 6 - len(password)
    return count





if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input().strip())

    password = input()

    answer = minimumNumber(n, password)

    fptr.write(str(answer) + '\n')

    fptr.close()

File: Problem Solving/Algorithms/Implementation/Strong Password.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'minimumNumber' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. STRING password
 */

int minimumNumber(int n, string password) {
    // Return the minimum number of characters to make the password strong
    string numbers = "0123456789";
    string lower_case = "abcdefghijklmnopqrstuvwxyz";
    string upper_case = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    string special_characters = "!@#$%^&*()-+";

    bool has_numbers = false, has_lower_case = false;
    bool has_upper_case = false, has_special_characters = false;
    
    for (auto ch: password) {
        if (numbers.find(ch) != numbers.npos)
            has_numbers = true;
        else if (lower_case.find(ch) != lower_case.npos)
            has_lower_case = true;
        else if (upper_case.find(ch) != upper_case.npos)
            has_upper_case = true;
        else if (special_characters.find(ch) != special_characters.npos)
            has_special_characters = true;
    }

    int count = 0;
    if (has_numbers == false) count++;
    if (has_lower_case == false) count++;
    if (has_upper_case == false) count++;
    if (has_special_characters == false) count++;

    if (password.size() + count < 6)
        return 6 - password.size();
    return count;



}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string n_temp;
    getline(cin, n_temp);

    int n = stoi(ltrim(rtrim(n_temp)));

    string password;
    getline(cin, password);

    int answer = minimumNumber(n, password);

    fout << answer << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}
